### PR TITLE
[core] Enforce a deadlock-free locking order for Mutexes.

### PR DESCRIPTION
**Pull request by jimblandy**
If `debug_assertions` or the `"validate-locks"` feature are enabled, change `wgpu-core` to use a wrapper around `parking_lot::Mutex` that checks for potential deadlocks.

### CODE REVIEW

**Comment by jimblandy** [comment]
This has already found a cycle, although it might take a lot of threads to hit it. If you uncomment out the indicated code in `wgpu-core/src/lock/rank.rs`, then you'll get this error mesage:

```
error[E0391]: cycle detected when const checking `lock::rank::COMMAND_BUFFER_DATA`
  --> wgpu-core/src/lock/rank.rs:28:35
   |
28 |                   followers: 0 $( | $follower.bit )*,
   |                                     ^^^^^^^^^
...
39 | / define_lock_ranks! {
40 | |     rank COMMAND_BUFFER_DATA 0 followed by {
41 | |         DEVICE_USAGE_SCOPES,
42 | |         SHARED_TRACKER_INDEX_ALLOCATOR_INNER,
...  |
83 | |     rank SURFACE_PRESENTATION 22 followed by { };
84 | | }
   | |_- in this macro invocation
   |
note: ...which requires const checking `lock::rank::BUFFER_MAP_STATE`...
  --> wgpu-core/src/lock/rank.rs:28:35
   |
28 |                   followers: 0 $( | $follower.bit )*,
   |                                     ^^^^^^^^^
...
39 | / define_lock_ranks! {
40 | |     rank COMMAND_BUFFER_DATA 0 followed by {
41 | |         DEVICE_USAGE_SCOPES,
42 | |         SHARED_TRACKER_INDEX_ALLOCATOR_INNER,
...  |
83 | |     rank SURFACE_PRESENTATION 22 followed by { };
84 | | }
   | |_- in this macro invocation
note: ...which requires const checking `lock::rank::DEVICE_PENDING_WRITES`...
  --> wgpu-core/src/lock/rank.rs:28:35
   |
28 |                   followers: 0 $( | $follower.bit )*,
   |                                     ^^^^^^^^^
...
39 | / define_lock_ranks! {
40 | |     rank COMMAND_BUFFER_DATA 0 followed by {
41 | |         DEVICE_USAGE_SCOPES,
42 | |         SHARED_TRACKER_INDEX_ALLOCATOR_INNER,
...  |
83 | |     rank SURFACE_PRESENTATION 22 followed by { };
84 | | }
   | |_- in this macro invocation
note: ...which requires const checking `lock::rank::DEVICE_LIFE_TRACKER`...
  --> wgpu-core/src/lock/rank.rs:28:35
   |
28 |                   followers: 0 $( | $follower.bit )*,
   |                                     ^^^^^^^^^
...
39 | / define_lock_ranks! {
40 | |     rank COMMAND_BUFFER_DATA 0 followed by {
41 | |         DEVICE_USAGE_SCOPES,
42 | |         SHARED_TRACKER_INDEX_ALLOCATOR_INNER,
...  |
83 | |     rank SURFACE_PRESENTATION 22 followed by { };
84 | | }
   | |_- in this macro invocation
note: ...which requires const checking `lock::rank::DEVICE_TEMP_SUSPECTED`...
  --> wgpu-core/src/lock/rank.rs:28:35
   |
28 |                   followers: 0 $( | $follower.bit )*,
   |                                     ^^^^^^^^^
...
39 | / define_lock_ranks! {
40 | |     rank COMMAND_BUFFER_DATA 0 followed by {
41 | |         DEVICE_USAGE_SCOPES,
42 | |         SHARED_TRACKER_INDEX_ALLOCATOR_INNER,
...  |
83 | |     rank SURFACE_PRESENTATION 22 followed by { };
84 | | }
   | |_- in this macro invocation
   = note: ...which again requires const checking `lock::rank::COMMAND_BUFFER_DATA`, completing the cycle
   = note: cycle used when running analysis passes on this crate
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
   = note: this error originates in the macro `define_lock_ranks` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by ErichDonGubler** [comment]
I'll take on review for this one. I've already done quite a bit of review with Jim as he's explained how the locking mechanism works. We'll both work to polish this up, and document and test properly. 🫡

File: wgpu-core/src/lock/validating.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/validating.rs]: - [x] issue: We should add some documentation to this symbol. At the very least, we should point folks towards this and the non-validating implementation being selected based on Cargo features, preferably by linking to the details you've already written elsewhere. Maybe something like this:

```suggestion
/// Mutual exclusion lock variant for resources in core which validates lock ordering. See also [`super`].
pub struct Mutex<T> {
    inner: parking_lot::Mutex<T>,
    rank: LockRank,
}
```

File: wgpu-core/src/lock/rank.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/rank.rs]: nitpick: RE: current commit contents ([0f1f20bbd55fb0f87d08b3d41f848cc2f9a392c8](https://github.com/gfx-rs/wgpu/commit/0f1f20bbd55fb0f87d08b3d41f848cc2f9a392c8)) and this DAG: You noted in first round of review on-call that writing this DAG was empirically driven. We should at least include that in the commit message (but maybe not a comment, since that seems both prone to being outdated and unimportant for actually working with it).

File: wgpu-core/src/lock/validating.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/validating.rs]: - [x] issue: I mention this elsewhere, but: we should add tests here. That can help folks wanting to understand this API much more easily, _and_ we should prove that we're actually preventing the deadlocks we say we are. Perhaps doc. tests would be appropriate for the most straightforward cases, with some `#[test]` functions for others?

    The only complication I can see here is that using something other than the current lock ranking requires an interchangeability of some kind, either via trait or a different value that can be passed in (maybe as a `cfg(test)`-only constructor?).

File: wgpu-core/src/lock/rank.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/rank.rs:1]: thought: RE: performance: I won't block this PR based on performance concerns, since we don't have any benchmarks. _But_, for the record, I expect that this would have been something that would have been nice to evaluate in terms of performance. I'll link to #5443 FTR.

File: wgpu-core/src/lock/mod.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/mod.rs]: nitpick: You don't need double-quotes inside the inline code block here. I understand that that's what somebody would literally write in a `features` array element in a `Cargo.toml`, but it's inconsistent with the styling we use at the top level of `wgpu` ([example](https://docs.rs/wgpu/0.19.3/wgpu/index.html)) and `wgpu-core` ([example](https://docs.rs/wgpu-core/0.19.3/wgpu_core/index.html)).

```suggestion
builds, and when the `validate-locks` feature is enabled.
```

File: wgpu-core/src/lock/mod.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/mod.rs]: nitpick: Links help us keep valid names, _and_ they're convenient for curious users. We should use them here.

TODO: suggestion

File: wgpu-core/src/lock/mod.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/mod.rs]: thought(non-blocking): You are the only person I know personally that prefers multi-line module commentary, as opposed to prepending `//!` to everything. 🤔

File: wgpu-core/src/lock/non_validating.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/non_validating.rs]: nitpick: It doesn't make sense to me to `allow(dead_code)` in a new API that only `wgpu-core` consumes. Is there something I'm missing?

**Comment by jimblandy** [comment]
One wrinkle: This has already detected a deadlock, and I haven't pushed very far yet, so I'd guess there'd be more to come. We'll have to fix them before we can actually land this.

[Edit: no, we can just make it only enabled when the feature is turned on, leaving both debug and release builds unaffected. This means we're not getting test coverage for the instrumented lock types, but at least we can have it in tree, rather than sitting on a branch as an indefinitely long list of deadlocks gets addressed.]

File: wgpu-core/src/lock/validating.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/validating.rs]: question: Is there a reason we can't assert before actually changing state? It's usually better to avoid invalid states entirely, but I'm not sure if it's reasonable to do here.

Comment (ErichDonGubler) [reply wgpu-core/src/lock/validating.rs]: - [x] ~~issue: We should test this case specifically!~~ EDIT: Invalidated.

File: wgpu-core/src/lock/validating.rs
Comment (ErichDonGubler) [reply wgpu-core/src/lock/validating.rs]: - [x] This feedback also applies to the `non_validating::Mutex` symbol, of course.

**Review by ErichDonGubler — COMMENTED** [review]
Not gonna commit to a review outcome yet, since this isn't out of draft yet. :^)

File: wgpu-core/src/lock/non_validating.rs
Comment (ErichDonGubler) [reply wgpu-core/src/lock/non_validating.rs]: Ah, you're doing this because this might not be used when not validating. Is there even a reason to compile the modules when one is not selected? At the very least, it seems we can make this suppression use some meaningful `cfg` conditions. Perhaps in the parent module declaration we could use `#[cfg_attr(…, allow(dead_code))]` instead to keep the mirrored conditions closer together?

File: wgpu-core/src/lock/validating.rs
Comment (jimblandy) [reply wgpu-core/src/lock/validating.rs]: I agree in general, but checking afterwards fits better with the `Cell` API, because we can do a single access with `replace`. My thought was, once a thread has panicked, its state doesn't matter much.

Comment (jimblandy) [reply wgpu-core/src/lock/validating.rs]: A test would need to check the thread's state after a panic, which doesn't seem that interesting.

File: wgpu-core/src/lock/mod.rs
Comment (jimblandy) [reply wgpu-core/src/lock/mod.rs]: I guess I wished I'd used it for the `wgpu-hal` module docs, because it would have made it easier to paste into the `README.md` file. (Yes, I know there is tooling for that.)

I've changed this to use `\\!`.

File: wgpu-core/src/lock/ranked.rs
Comment (cwfitzgerald) [inline wgpu-core/src/lock/ranked.rs]:
```suggestion
//! Thus, the system as a whole is always able to make progress: it is free of
```

**Review by cwfitzgerald — APPROVED** [review]
This is absolutely amazing stuff!

Have a docs nit, but I love the docs, they are _very_ clear

**Comment by jimblandy** [comment]
@cwfitzgerald Thanks for the review! @ErichDonGubler and I agreed this should have unit tests, so I'll work on that tomorrow and then ask Erich to take another look.

**Comment by cwfitzgerald** [comment]
Sounds good, could @ErichDonGubler just mark it as request changes so I don't accidentally merge something I shouldn't :)

**Review by ErichDonGubler — CHANGES_REQUESTED** [review]
Just waiting on feedback I've already left. 😊

File: wgpu-core/src/lock/validating.rs
Comment (jimblandy) [reply wgpu-core/src/lock/validating.rs]: All `Mutex` and guard types now have docs, and there's an explanation of how we select an implementation.

File: wgpu-core/src/lock/rank.rs
Comment (jimblandy) [reply wgpu-core/src/lock/rank.rs]: I've filed #5572 to track filling the ranking out.

Comment (jimblandy) [reply wgpu-core/src/lock/rank.rs]: I've mentioned #5572 in the commit message.

File: wgpu-core/src/lock/validating.rs
Comment (jimblandy) [reply wgpu-core/src/lock/validating.rs]: All these have docs now.

File: wgpu-core/src/lock/validating.rs
Comment (jimblandy) [reply wgpu-core/src/lock/validating.rs]: Okay, I think I've got some decent tests now.

**Comment by jimblandy** [comment]
@ErichDonGubler, I took most of your suggestions but you had one that I didn't want to take. I've explained and left it unresolved; let me know what you think.

File: wgpu-core/src/lock/mod.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/mod.rs:29]: nitpick: We're still unconditionally using `allow(dead_code)` in the `rank` module. I think we should have a similar `#[cfg_attr(feature = "validate-locks", allow(dead_code))]` here instead?

File: wgpu-core/src/lock/ranked.rs
Comment (ErichDonGubler) [inline wgpu-core/src/lock/ranked.rs]: nitpick(non-blocking): Linking to `parking_lot::Mutex` plz? 😀

EDIT: This also applies to `MutexGuard`.

**Review by ErichDonGubler — APPROVED** [review]
Only nitpick feedback left. Approving to enable progress, since I trust @jimblandy to duly consider the feedback I've left.

File: wgpu-core/src/lock/mod.rs
Comment (jimblandy) [reply wgpu-core/src/lock/mod.rs:29]: Hmm, it's funny, it's only the fields of `LockRank` that aren't read. We certainly use the rank constants unconditionally in `wgpu-core`.

Comment (jimblandy) [reply wgpu-core/src/lock/mod.rs:29]: It turns out that... the `cfg_attr` conditions were swapped??? Anyway, now that they're fixed, there are no extraneous `allow`s in `wgpu-core/src/lock`.

**Comment by jimblandy** [comment]
@ErichDonGubler Hmm. Plot twist.

1) Just today I remarked on Rust's wisdom in flagging the use of `feature = "foo"` in the code when there's no `"foo"` feature in `Cargo.toml`. This is completely untrue: rustc just treats them as if they're not enabled.

2) We want our CI to run tests with `--all-features`, which would enable `"validate-locks"`, which causes a panic in ordinary operation.

I'm not sure what the best way forward is.

- The easiest is to omit the feature from `Cargo.toml`. If people want to experiment it, they'll have to add it back themselves.

- We could just comment it out, but that has the disadvantage that it's not clear what you need to uncomment to make things work. The nice thing about putting `feature = "validate-locks"`, even without a `Cargo.toml` entry, is that its use marks everything that needs to be adjusted, and it's pretty obvious what to do if you want to try it out.

Other ideas??

**Comment by jimblandy** [comment]
In chat we decided to go with `cfg(validate_locks)`.

Edit: Err, cfgs are compilation-wide, so actually it's `cfg(wgpu_validate_locks)`.

**Pull request merged**
